docs/imds_retry_based_on_errors.md

```diff
@@ -0,0 +1,65 @@
+# MSAL & MSI IMDS Error Handling and Retry Strategy Specification
+
+## Overview
+This document defines the error handling and retry strategy for MSAL when interacting with the IMDS (Instance Metadata Service) endpoint for Managed Identity (MSI) token acquisition.
+
+---
+
+## 1️⃣ HTTP Status Codes & Recommended Actions
+
+| **HTTP Status Code** | **Error Reason**                               | **Recommended Action**                      | **Retry Delay Strategy**                 |
+|----------------------|-----------------------------------------------|---------------------------------------------|-----------------------------------------|
+| **400**             | Bad Request (Invalid Parameters)               | **Do not retry**, fix request               | **No retry**                             |
+| **401**             | Unauthorized                                  | **Do not retry**, check authentication setup | **No retry**                             |
+| **403**             | Forbidden                                     | **Do not retry**, verify permissions       | **No retry**                             |
+| **404**             | IMDS endpoint is updating / Identity Not Found | Retry with Exponential Backoff (max 5 retries) | **1s → 2s → 4s → 8s → 16s**            |
+| **408**             | Request Timeout                                | Retry with Exponential Backoff (max 5 retries) | **1s → 2s → 4s → 8s → 16s (max 16s)**            |
+| **410**             | IMDS is undergoing updates                    | Wait up to **70 seconds**, then retry      | **70s (fixed wait)**                    |
```

Comment:
70s is a really really long time. it may be taht IMDS finishes the update sooner. Can we do exponential here as well?

Reply:
Based on Nidhi's comment, "410 is really very rare. and it is recommended by the IMDS team", read more [here](https://eng.ms/docs/cloud-ai-platform/azure-core/core-compute-and-host/general-purpose-host-arunki/azure-instance-metadata-service/compute-azlinux-metadataserver/troubleshooting/unable-to-reach-imds#mitigate-http-status-code-410)

Reply:
I also think 70 seconds is too long. However I don't mind implementing it if Bogdan gives the ok.

Reply:
@bgavrilMS Maybe we can log a message to the user on a 410 error before waiting 70 seconds?

Reply:
As Nidhi mentions this doesn't happen frequently, maybe we can fine tune based on telemetry?

Reply:
> @bgavrilMS Maybe we can log a message to the user on a 410 error before waiting 70 seconds?

We should have detailed logging statements for each of these retries. Logs are for us (the dev team) to troubleshoot.

Reply:
Are you folks ok with the spec? can we merge and start work on this? @bgavrilMS @Robbie-Microsoft 

If not, I can start a conversation with Nidhi and us.

Reply:
Well, no. We still need to address Bogdan's comment. For 410,  should we check every 10 seconds for up to 70 seconds?

Reply:
Updated 410 to use exponential retry with 4 attempts max

Reply:
@Robbie-Microsoft I added you to a thread with Nidhi, updated the spec based on her comment. Once you ack this. I can close the comment.

Reply:
Retry every 10 seconds for a maximum of 70 seconds. Log a statement every retry.